### ISSUE TITLE
DataContractJsonSerializerSettings.UseSimpleDictionaryFormat bug

### ISSUE THREAD

**Issue by durswd** [mono/mono#7717]
## Steps to Reproduce

Run these codes.
(Refered this website http://yasuand.hatenablog.com/entry/2013/09/12/051655)

```csharp
using System;
using System.Collections.Generic;
using System.Globalization;
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Json;
using System.Text;

[DataContract]
class Data
{
    [DataMember]
    public Dictionary<string, double> xs;
}

class Program
{
    static void Main()
    {
        var json = @"{""xs"":{""1"":1, ""e"":2.718, ""pi"":3.14}}";
        var settings = new DataContractJsonSerializerSettings();
        settings.UseSimpleDictionaryFormat = true;
        var serializer = new DataContractJsonSerializer(typeof(Data), settings);
        using (var ms = new MemoryStream(Encoding.UTF8.GetBytes(json)))
        {
            var data = (Data)serializer.ReadObject(ms);
            foreach (var x in data.xs)
            {
                Console.WriteLine("{0}\t{1}", x.Key, x.Value);
            }
        }
    }
}
```

## Current Behavior

System.Reflection.TargetInvocationException is thrown in 

System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadSimpleDictionary(System.Runtime.Serialization.CollectionDataContract collectionContract, System.RuntimeType keyValueType) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:522

## Expected Behavior

Run without an exception.

### On which platforms did you notice this

macOS

**Version Used**:

Mono JIT compiler version 5.8.0.129 (2017-10/835a98d8559 Fri Feb 23 21:09:37 EST 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
	TLS:           normal
	SIGSEGV:       altstack
	Notification:  kqueue
	Architecture:  amd64
	Disabled:      none
	Misc:          softdebug 
	LLVM:          yes(3.6.0svn-mono-master/8b1520c8aae)
	GC:            sgen (concurrent by default)

### Stacktrace

```
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadSimpleDictionary(System.Runtime.Serialization.CollectionDataContract collectionContract, System.RuntimeType keyValueType) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:522
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadCollection(System.Runtime.Serialization.CollectionDataContract collectionContract) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:423
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadCollectionFromJson(System.Runtime.Serialization.Json.JsonReaderDelegator xmlReader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context, System.Xml.XmlDictionaryString emptyDictionaryString, System.Xml.XmlDictionaryString itemName, System.Runtime.Serialization.CollectionDataContract collectionContract) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:110
System.Runtime.Serialization.Json.JsonFormatReaderGenerator.CriticalHelper.(System.Runtime.Serialization.Json.JsonReaderDelegator xr, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson ctx, System.Xml.XmlDictionaryString emptyDS, System.Xml.XmlDictionaryString inm, System.Runtime.Serialization.CollectionDataContract cc) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:21
System.Runtime.Serialization.Json.JsonCollectionDataContract.ReadJsonValueCore(System.Runtime.Serialization.Json.JsonReaderDelegator jsonReader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/JsonCollectionDataContract.cs:129
System.Runtime.Serialization.Json.JsonDataContract.ReadJsonValue(System.Runtime.Serialization.Json.JsonReaderDelegator jsonReader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/JsonDataContract.cs:84
System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadJsonValue(System.Runtime.Serialization.CollectionDataContract contract, System.Runtime.Serialization.Json.JsonReaderDelegator reader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs:376
System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson.ReadDataContractValue(System.Runtime.Serialization.CollectionDataContract dataContract, System.Runtime.Serialization.Json.JsonReaderDelegator reader) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/XmlObjectSerializerReadContextComplexJson.cs:134
System.Runtime.Serialization.XmlObjectSerializerReadContext.InternalDeserialize(System.Runtime.Serialization.Json.JsonReaderDelegator reader, string name, string ns, System.RuntimeType declaredType, System.Runtime.Serialization.CollectionDataContract dataContract) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializerReadContext.cs:233
System.Runtime.Serialization.XmlObjectSerializerReadContext.InternalDeserialize(System.Runtime.Serialization.Json.JsonReaderDelegator xmlReader, int id, System.RuntimeTypeHandle declaredTypeHandle, string name, string ns) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializerReadContext.cs:131
System.Runtime.Serialization.XmlObjectSerializerReadContextComplex.InternalDeserialize(System.Runtime.Serialization.Json.JsonReaderDelegator xmlReader, int declaredTypeID, System.RuntimeTypeHandle declaredTypeHandle, string name, string ns) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializerReadContextComplex.cs:114
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.InternalDeserialize(System.RuntimeType type, string name) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:357
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadValue(System.RuntimeType type, string name) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:349
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadMembers(int index, System.Runtime.Serialization.ClassDataContract classContract, System.Runtime.Serialization.BitFlagsGenerator expectedElements, int memberIndex) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:248
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadMembers(System.Runtime.Serialization.ClassDataContract classContract, System.Runtime.Serialization.ExtensionDataObject extensionData) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:223
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadClass(System.Runtime.Serialization.ClassDataContract classContract) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:202
System.Runtime.Serialization.Json.JsonFormatReaderInterpreter.ReadFromJson(System.Runtime.Serialization.Json.JsonReaderDelegator xmlReader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context, System.Xml.XmlDictionaryString emptyDictionaryString, System.Xml.XmlDictionaryString[] memberNames) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:80
System.Runtime.Serialization.Json.JsonFormatReaderGenerator.CriticalHelper.(System.Runtime.Serialization.Json.JsonReaderDelegator xr, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson ctx, System.Xml.XmlDictionaryString emptyDictionaryString, System.Xml.XmlDictionaryString[] memberNames) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs:16
System.Runtime.Serialization.Json.JsonClassDataContract.ReadJsonValueCore(System.Runtime.Serialization.Json.JsonReaderDelegator jsonReader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/JsonClassDataContract.cs:110
System.Runtime.Serialization.Json.JsonDataContract.ReadJsonValue(System.Runtime.Serialization.Json.JsonReaderDelegator jsonReader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/JsonDataContract.cs:84
System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadJsonValue(System.Runtime.Serialization.ClassDataContract contract, System.Runtime.Serialization.Json.JsonReaderDelegator reader, System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson context) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs:376
System.Runtime.Serialization.Json.XmlObjectSerializerReadContextComplexJson.ReadDataContractValue(System.Runtime.Serialization.ClassDataContract dataContract, System.Runtime.Serialization.Json.JsonReaderDelegator reader) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/XmlObjectSerializerReadContextComplexJson.cs:134
System.Runtime.Serialization.XmlObjectSerializerReadContext.InternalDeserialize(System.Runtime.Serialization.Json.JsonReaderDelegator reader, string name, string ns, System.RuntimeType declaredType, System.Runtime.Serialization.ClassDataContract dataContract) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializerReadContext.cs:233
System.Runtime.Serialization.XmlObjectSerializerReadContext.InternalDeserialize(System.Runtime.Serialization.Json.JsonReaderDelegator xmlReader, System.RuntimeType declaredType, System.Runtime.Serialization.ClassDataContract dataContract, string name, string ns) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializerReadContext.cs:144
System.Runtime.Serialization.XmlObjectSerializerReadContextComplex.InternalDeserialize(System.Runtime.Serialization.Json.JsonReaderDelegator xmlReader, System.RuntimeType declaredType, System.Runtime.Serialization.ClassDataContract dataContract, string name, string ns) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializerReadContextComplex.cs:144
System.Runtime.Serialization.Json.DataContractJsonSerializer.InternalReadObject(System.Runtime.Serialization.Json.JsonReaderDelegator xmlReader, bool verifyObjectName) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs:435
System.Runtime.Serialization.XmlObjectSerializer.InternalReadObject(System.Runtime.Serialization.Json.JsonReaderDelegator reader, bool verifyObjectName, System.Runtime.Serialization.DataContractResolver dataContractResolver) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializer.cs:283
System.Runtime.Serialization.XmlObjectSerializer.ReadObjectHandleExceptions(System.Runtime.Serialization.Json.JsonReaderDelegator reader, bool verifyObjectName, System.Runtime.Serialization.DataContractResolver dataContractResolver) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializer.cs:313
System.Runtime.Serialization.XmlObjectSerializer.ReadObjectHandleExceptions(System.Runtime.Serialization.Json.JsonReaderDelegator reader, bool verifyObjectName) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/XmlObjectSerializer.cs:294
System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.Runtime.Serialization.Json.XmlJsonReader reader) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs:258
System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.IO.MemoryStream stream) in /Users/builder/data/lanes/5533/mono-mac-sdk/external/bockbuild/builds/mono-x64/mcs/class/referencesource/System.Runtime.Serialization/System/Runtime/Serialization/Json/DataContractJsonSerializer.cs:243
Program.Main() in /Users/xxx/jsonTest/jsonTest/Program.cs:26
```

**Comment by akoeplinger** [comment]
I can reproduce this with mono master/081c79e687131e2b26cd05ca28c20aa3cf65c4e0

**Comment by marek-safar** [comment]
This is because at https://github.com/mono/mono/blob/7dea71ce64bb77cb96ff3dfb06bb9141fb73c3f0/mcs/class/System.Runtime.Serialization/ReferenceSources/JsonFormatReaderGenerator_static.cs#L511 we don't handle keyParseMode of `AsString` and key is set remains set to `null`

**Comment by galilov** [comment]
Is there any workaround? I faced this issue

**Comment by NextTurn** [comment]
This issue was fixed in #9342 and the changes were shipped in mono-5.18.0.225

Sorry for keeping you waiting :-)
